File: app/modules/sticky_sessions/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field, model_validator

from app.db.models import StickySessionKind
from app.modules.shared.schemas import DashboardModel

StickySessionSortBy = Literal["updated_at", "created_at", "account", "key"]
StickySessionSortDir = Literal["asc", "desc"]


class StickySessionEntryResponse(DashboardModel):
    key: str
    display_name: str
    kind: StickySessionKind
    created_at: datetime
    updated_at: datetime
    expires_at: datetime | None = None
    is_stale: bool


class StickySessionsListResponse(DashboardModel):
    entries: list[StickySessionEntryResponse] = Field(default_factory=list)
    stale_prompt_cache_count: int = 0
    total: int = 0
    has_more: bool = False


class StickySessionIdentifier(DashboardModel):
    key: str = Field(min_length=1)
    kind: StickySessionKind


class StickySessionDeleteResponse(DashboardModel):
    status: str


class StickySessionsDeleteRequest(DashboardModel):
    sessions: list[StickySessionIdentifier] = Field(min_length=1, max_length=500)

    @model_validator(mode="after")
    def validate_unique_sessions(self) -> "StickySessionsDeleteRequest":
        seen: set[tuple[str, StickySessionKind]] = set()
        for session in self.sessions:
            target = (session.key, session.kind)
            if target in seen:
                raise ValueError("duplicate sticky session targets are not allowed")
            seen.add(target)
        return self


class StickySessionDeleteFailure(DashboardModel):
    key: str
    kind: StickySessionKind
    reason: str


class StickySessionsDeleteResponse(DashboardModel):
    deleted_count: int
    deleted: list[StickySessionIdentifier] = Field(default_factory=list)
    failed: list[StickySessionDeleteFailure] = Field(default_factory=list)


class StickySessionsDeleteFilteredRequest(DashboardModel):
    stale_only: bool = False
    account_query: str = ""
    key_query: str = ""


class StickySessionsDeleteFilteredResponse(DashboardModel):
    deleted_count: int


class StickySessionsPurgeRequest(DashboardModel):
    stale_only: Literal[True] = True


class StickySessionsPurgeResponse(DashboardModel):
    deleted_count: int

File: frontend/src/features/sticky-sessions/schemas.ts
import { z } from "zod";

export const STICKY_SESSION_KINDS = ["codex_session", "sticky_thread", "prompt_cache"] as const;
export const STICKY_SESSION_SORT_FIELDS = ["updated_at", "created_at", "account", "key"] as const;
export const STICKY_SESSION_SORT_DIRECTIONS = ["asc", "desc"] as const;

export const StickySessionKindSchema = z.enum(STICKY_SESSION_KINDS);
export const StickySessionSortBySchema = z.enum(STICKY_SESSION_SORT_FIELDS);
export const StickySessionSortDirSchema = z.enum(STICKY_SESSION_SORT_DIRECTIONS);

export const StickySessionEntrySchema = z.object({
  key: z.string().min(1),
  displayName: z.string().min(1),
  kind: StickySessionKindSchema,
  createdAt: z.string().datetime({ offset: true }),
  updatedAt: z.string().datetime({ offset: true }),
  expiresAt: z.string().datetime({ offset: true }).nullable(),
  isStale: z.boolean(),
});

export const StickySessionIdentifierSchema = z.object({
  key: z.string().min(1),
  kind: StickySessionKindSchema,
});

export const StickySessionsDeleteRequestSchema = z.object({
  sessions: z
    .array(StickySessionIdentifierSchema)
    .min(1)
    .max(500)
    .refine(
      (sessions) => new Set(sessions.map((session) => `${session.kind}:${session.key}`)).size === sessions.length,
      "Duplicate sticky session targets are not allowed",
    ),
});

export const StickySessionDeleteFailureSchema = z.object({
  key: z.string().min(1),
  kind: StickySessionKindSchema,
  reason: z.string().min(1),
});

export const StickySessionsListResponseSchema = z.object({
  entries: z.array(StickySessionEntrySchema).default([]),
  stalePromptCacheCount: z.number().int().nonnegative().default(0),
  total: z.number().int().nonnegative().default(0),
  hasMore: z.boolean().default(false),
});

export const StickySessionsListParamsSchema = z.object({
  staleOnly: z.boolean().default(false),
  accountQuery: z.string().default(""),
  keyQuery: z.string().default(""),
  sortBy: StickySessionSortBySchema.default("updated_at"),
  sortDir: StickySessionSortDirSchema.default("desc"),
  offset: z.number().int().nonnegative().default(0),
  limit: z.number().int().positive().max(500).default(10),
});

export const StickySessionsDeleteResponseSchema = z.object({
  deletedCount: z.number().int().nonnegative(),
  deleted: z.array(StickySessionIdentifierSchema).default([]),
  failed: z.array(StickySessionDeleteFailureSchema).default([]),
});

export const StickySessionsDeleteFilteredRequestSchema = z.object({
  staleOnly: z.boolean().default(false),
  accountQuery: z.string().default(""),
  keyQuery: z.string().default(""),
});

export const StickySessionsDeleteFilteredResponseSchema = z.object({
  deletedCount: z.number().int().nonnegative(),
});

export const StickySessionsPurgeRequestSchema = z.object({
  staleOnly: z.boolean().default(true),
});

export const StickySessionsPurgeResponseSchema = z.object({
  deletedCount: z.number().int().nonnegative(),
});

export type StickySessionKind = z.infer<typeof StickySessionKindSchema>;
export type StickySessionSortBy = z.infer<typeof StickySessionSortBySchema>;
export type StickySessionSortDir = z.infer<typeof StickySessionSortDirSchema>;
export type StickySessionEntry = z.infer<typeof StickySessionEntrySchema>;
export type StickySessionIdentifier = z.infer<typeof StickySessionIdentifierSchema>;
export type StickySessionsDeleteRequest = z.infer<typeof StickySessionsDeleteRequestSchema>;
export type StickySessionsListResponse = z.infer<typeof StickySessionsListResponseSchema>;
export type StickySessionsListParams = z.infer<typeof StickySessionsListParamsSchema>;
export type StickySessionDeleteFailure = z.infer<typeof StickySessionDeleteFailureSchema>;
export type StickySessionsDeleteResponse = z.infer<typeof StickySessionsDeleteResponseSchema>;
export type StickySessionsDeleteFilteredRequest = z.infer<typeof StickySessionsDeleteFilteredRequestSchema>;
export type StickySessionsDeleteFilteredResponse = z.infer<typeof StickySessionsDeleteFilteredResponseSchema>;
export type StickySessionsPurgeRequest = z.infer<typeof StickySessionsPurgeRequestSchema>;
export type StickySessionsPurgeResponse = z.infer<typeof StickySessionsPurgeResponseSchema>;
